cli/core/test/e2e/framework/plugin_operations.go

```diff
@@ -0,0 +1,267 @@
+type PluginOps interface {
+
+	// GenerateScriptBasedPluginBinaries generates script based plugin binaries for given plugin metadata
+	GenerateScriptBasedPluginBinaries(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBinary publishes the plugin binaries to given registry bucket and returns the plugin distribution urls
+	PublishPluginBinary(pluginsInfo []*PluginMeta) (discoveryUrls []string, errs []error)
+
+	// GeneratePluginBundle generates plugin bundle in local file system for given plugin metadata
+	GeneratePluginBundle(pluginsMeta []*PluginMeta) ([]string, []error)
+
+	// PublishPluginBundle publishes the plugin bundles to given registry bucket and returns the plugins discovery urls
+	PublishPluginBundle(pluginsInfo []*PluginMeta) (discoveryUrls []string, errs []error)
+}
+
+// CLIPlugin for plugin overlay info
+type CLIPlugin struct {
+	metav1.TypeMeta `yaml:",inline"`
+	Metadata        metav1.ObjectMeta         `json:"metadata"`
+	Spec            plugintypes.CLIPluginSpec `json:"spec"`
+}
+
+// localOCIPluginOps is the implementation of PluginOps interface
+type localOCIPluginOps struct {
+	PluginOps
+	cmdExe    CmdOps
+	registry  PluginRegistry
+	imgpkgOps ImgpkgOps
+}
+
+func NewLocalOCIPluginOps(registry PluginRegistry) PluginOps {
+	return &localOCIPluginOps{
+		cmdExe:    NewCmdOps(),
+		registry:  registry,
+		imgpkgOps: NewImgpkgOps(),
+	}
+}
+
+func (po *localOCIPluginOps) GenerateScriptBasedPluginBinaries(pluginsMeta []*PluginMeta) ([]string, []error) {
+	pluginsProcessed := make(map[string]bool)
+	size := len(pluginsMeta)
+	pluginBinaryFilePaths := make([]string, size)
+	errs := make([]error, size)
+	for i, pm := range pluginsMeta {
+		if _, exists := pluginsProcessed[pm.name]; exists {
+			errs[i] = errors.New("plugin name already exists, currently multiple versions of same plugin not supported")
+			continue
+		}
+		pluginsProcessed[pm.name] = true
+
+		// Set plugin local dir path if not, to generate binary image and bundle to publish to registry
+		if pm.pluginLocalPath == "" {
+			pm.pluginLocalPath = filepath.Join(TestPluginsDirPath, pm.name)
+		}
+
+		pluginBinaryFilePath, err := po.generatePluginBinary(pm)
+		if err != nil {
+			errs[i] = err
+			continue
+		}
+		pm.pluginBinaryFilePath = pluginBinaryFilePath
+		pluginBinaryFilePaths[i] = pm.pluginBinaryFilePath
+	}
+	return pluginBinaryFilePaths, errs
+}
+
+// PublishPluginBinary publishes the plugin binaries to given registry bucket and returns the plugin distribution urls
+func (po *localOCIPluginOps) PublishPluginBinary(pluginsMeta []*PluginMeta) (discoveryUrls []string, errs []error) {
+	size := len(pluginsMeta)
+	distributionUrls := make([]string, size)
+	errs = make([]error, size)
+	for i, pm := range pluginsMeta {
+		// Set registry discovery url if not set already
+		if pm.registryDiscoveryURL == "" {
+			pm.registryDiscoveryURL = filepath.Join(po.registry.GetRegistryURLWithDefaultCLIPluginsBucket(), ("/" + pm.name + "/"))
```

Comment:
Need to incorporate `target` information here to uniquely publish plugin binaries.

Reply:
Done